#!/bin/bash -l
#PBS -l walltime=12:00:00
#PBS -l mem=100gb
#PBS -l nodes=1:ppn=20
#PBS -M MAIL 
#PBS -m aeb
#PBS -N bowtie_mapping
#PBS -A default_project
##[VERSION] pika 16.08
##[HELP] This tool will map a single sample to the refrence genome
##[HELP] This script is multithreaded
##[HELP] This script runs around 2 minute for 1M reads
##[HELP] In this script the sample, and flash options need to be changed
##
##[HOWTO] #Bowtie
##[HOWTO] #The sample, flash and genome directory needs to be changed
##[HOWTO] #start the module:
##[HOWTO] qsub bowtie_mapping.pbs

##[HELP] Options:
##[HELP] genome: the genome to use (abriviation if known or full path)
##[OPTIONS] genome	mandatory	sed 's:GENOME_DIR=\"\":GENOME_DIR=\"value\":g'
##[HELP] flash: use regular files (NO), only flashed (FLASH) or flashed and unflashed (BOTH)
##[OPTIONS] flash	optional	sed "s:FLASH=\"NO\":FLASH=\"value\":g"
##[HELP] bowtie_options: change the options of the mapper (standard: --local)
##[OPTIONS] bowtie_options	optional	sed "s:BOWTIE_OPTIONS=\"--local\":BOWTIE_OPTIONS=\"value\":g"



#loading the modules
#extra_modules
module load Bowtie2
module load SAMtools
module load picard

#setting all parameters (these could be changed)
PROJECT_DIR="";
GENOME_DIR="";
SAMPLE_DIR="$PROJECT_DIR/raw";
OUTPUT_DIR="$PROJECT_DIR/mapped";
#this can be NO (for regular files), FLASH (for only flashed data) or BOTH (for both flashed and unflashed data)
FLASH="NO";
SCRATCH_DIR=~;
SAMPLE="samplename";
THREADS="20";
#options can be --local or --end-to-end (more options see bowtie2 manual)
#local is ideal for genome data, end is ideal for gbs, quantseq, ...
BOWTIE_OPTIONS="--local";

if [ "$FLASH" -ne "NO" ];
then
	SAMPLE_DIR="$PROJECT_DIR/flash";
fi

#the actual script
mkdir -p $SCRATCH_DIR/genome;
cd $SCRATCH_DIR/genome;
rsync -ahrL $GENOME_DIR/* .;

GENOME_DIR="$SCRATCH_DIR/genome/genome";

#do the copy and the mapping
cd $SCRATCH_DIR;
TMPDIR=$SCRATCH_DIR;
TMP_DIR=`mktemp -d -t tmp.XXXXXXXX`;
cd $TMP_DIR;

rsync -ahrL $SAMPLE_DIR/$SAMPLE*.fastq.gz .;
gunzip *gz;
FASTQ_FILES="";
case "$FLASH" in
	"NO")
		#no flash files
		if [ -f $SAMPLE.R2.fastq ];
		then 
			#paired end mapping
			FASTQ_FILES="-1 $TMP_DIR/$SAMPLE.R1.fastq -2 $TMP_DIR/$SAMPLE.R2.fastq";
		else
			#single read mapping
			FASTQ_FILES="$TMP_DIR/$SAMPLE.R1.fastq";
		fi
	;;
	"FLASH")
		FASTQ_FILES="$TMP_DIR/$SAMPLE.extendedFrags.fastq";
	;;
	"BOTH")
		FASTQ_FILES="$TMP_DIR/$SAMPLE.extendedFrags.fastq";
		FASTQ_FILES_SECOND_BATCH="-1 $TMP_DIR/$SAMPLE.notCombined_1.fastq -2 $TMP_DIR/$SAMPLE.notCombined_2.fastq";
	;;
esac

bowtie2 -q -p $THREADS --seed 1 $BOWTIE_OPTIONS -x $GENOME_DIR $FASTQ_FILES > $TMP_DIR/mapped.sam 2> $TMP_DIR/$SAMPLE.bowtie2.log;
if [ "$FLASH" == "BOTH" ];
then
	bowtie2 -q -p $THREADS --seed 1 $BOWTIE_OPTIONS -x $GENOME_DIR $FASTQ_FILES_SECOND_BATCH > $TMP_DIR/mapped.notCombined.sam 2> $TMP_DIR/$SAMPLE.bowtie2.notCombined.log;
fi
rm *fastq;

samtools view -bS -@ $THREADS $TMP_DIR/mapped.sam > $TMP_DIR/mapped.bam;
rm -r $TMP_DIR/mapped.sam;
samtools sort -@ $THREADS $TMP_DIR/mapped.bam $TMP_DIR/$SAMPLE.sorted;
#samtools index -b $TMP_DIR/$SAMPLE.bam;
rm -r $TMP_DIR/mapped.bam;

if [ "$FLASH" == "BOTH" ];
then
	#sort second sam file
	mv $TMP_DIR/$SAMPLE.sorted.bam $TMP_DIR/$SAMPLE.flash.bam;
	rm $TMP_DIR/mapped.bam;
	samtools view -bS -@ $THREADS $TMP_DIR/mapped.notCombined.sam > $TMP_DIR/mapped.bam;
	rm $TMP_DIR/mapped.notCombined.sam;
	samtools sort -@ $THREADS $TMP_DIR/mapped.bam $TMP_DIR/$SAMPLE.notflash;
	rm $TMP_DIR/mapped.bam;
	#samtools index -b $TMP_DIR/$SAMPLE.notflash.bam;
	#merge the 2 bam files
	samtools view -H $TMP_DIR/$SAMPLE.notflash.bam | grep -v '^\@RG' > $TMP_DIR/header.sam;
	samtools view -H $TMP_DIR/$SAMPLE.notflash.bam | grep '^\@RG' | sort -u > $TMP_DIR/rg.sam;
	samtools view -H $TMP_DIR/$SAMPLE.flash.bam | grep '^\@RG' | sort -u >> $TMP_DIR/rg.sam;
	cat $TMP_DIR/rg.sam | sort -u >> $TMP_DIR/header.sam;
	samtools merge -h $TMP_DIR/header.sam $TMP_DIR/merged.bam $TMP_DIR/$SAMPLE.flash.bam $TMP_DIR/$SAMPLE.notflash.bam;
	rm $TMP_DIR/header.sam $TMP_DIR/rg.sam;
	rm $TMP_DIR/$SAMPLE.flash.bam;
	rm $TMP_DIR/$SAMPLE.notflash.bam;
	samtools sort -@ $THREADS $TMP_DIR/merged.bam $TMP_DIR/$SAMPLE.sorted;
	rm $TMP_DIR/merged.bam;
fi

#run picard tools for correcting readgroups...
picard AddOrReplaceReadGroups I=$TMP_DIR/$SAMPLE.sorted.bam O=$TMP_DIR/$SAMPLE.bam RGID=$SAMPLE RGLB=$SAMPLE RGPL=illumina RGSM=$SAMPLE RGPU=$SAMPLE RGCN=GenomicsCoreLeuven CREATE_INDEX=TRUE;

rm $TMP_DIR/$SAMPLE.sorted.bam;

mkdir -p $OUTPUT_DIR;
cd $OUTPUT_DIR;
rsync -ahr $TMP_DIR/$SAMPLE* .;
rm -rf $TMP_DIR/$SAMPLE;
rm -rf $SCRATCH_DIR/genome;


