#!/bin/bash -l
#PBS -l walltime=12:00:00
#PBS -l mem=100gb
#PBS -l nodes=1:ppn=20
#PBS -M MAIL
#PBS -m aeb
#PBS -N bwa_mapping_population
#PBS -A default_project
#This is a worker script: it uses the SAMPLE parameters
##[VERSION] pika 16.08
##[HELP] This is a worker script: it uses SAMPLE parameters
##[HELP] This tool will map the reads to the reference genome
##[HELP] This script will use only one cpu per sample
##[HELP] This script runs only 2 minutes per 1M reads
##[HELP] This scripts needs to be edited: change the flash parameter, and the genome directory in the prolog file
##
##[HOWTO] #BWA mapping for populations
##[HOWTO] #Create a data file (samples.txt), with a SAMPLE column
##[HOWTO] #This uses the 2015a (there can be a more recent worker installed)
##[HOWTO] module load worker/1.6.4-intel-2015a
##[HOWTO] #The script need to be edited: change the flash parameter
##[HOWTO] #The genome directory in the prolog file will need to be changed
##[HOWTO] #start the module:
##[HOWTO] wsub -batch bwa_mapping_population.pbs -data samples.txt -prolog bwa_mapping_population.prolog.sh
##[HOWTO] #When the THREADS variable in this script is changed, this can be ran as multithreaded population mapping, start like this (replace 20 with the number of given threads):
##[HOWTO] wsub -batch bwa_mapping_population.pbs -data samples.txt -prolog bwa_mapping_population.prolog.sh -threaded 20
#loading the modules
module use /apps/leuven/thinking/2015a/modules/all 
module use /staging/leuven/stg_00019/software/modulefiles
module load BWA/0.7.12-intel-2015a
module load SAMtools/1.1-intel-2015a
module load picard/2.2.2-Java1.8.0_77

#setting all parameters (these could be changed)
PROJECT_DIR="";
GENOME_DIR="$VSC_SCRATCH_NODE/genome/genome.fa";
SAMPLE_DIR="$PROJECT_DIR/raw";
OUTPUT_DIR="$PROJECT_DIR/mapped";
#this can be NO (for regular files), FLASH (for only flashed data) or BOTH (for both flashed and unflashed data)
FLASH="NO";
SCRATCH_DIR=$VSC_SCRATCH_NODE;
#This parameter can be changed for multithreaded mapping
THREADS="1";


#the actual script
mkdir -p  $SCRATCH_DIR/$SAMPLE;
cd $SCRATCH_DIR/$SAMPLE;

rsync -ahrL $SAMPLE_DIR/$SAMPLE*.fastq.gz $SCRATCH_DIR/$SAMPLE/;
gunzip *gz;
FASTQ_FILES="";
case "$FLASH" in
	"NO")
		#no flash files
		if [ -f $SAMPLE.R2.fastq ];
		then 
			#paired end mapping
			FASTQ_FILES="$SCRATCH_DIR/$SAMPLE/$SAMPLE.R1.fastq $SCRATCH_DIR/$SAMPLE/$SAMPLE.R2.fastq";
		else
			#single read mapping
			FASTQ_FILES="$SCRATCH_DIR/$SAMPLE/$SAMPLE.R1.fastq";
		fi
	;;
	"FLASH")
		FASTQ_FILES="$SCRATCH_DIR/$SAMPLE/$SAMPLE.extendedFrags.fastq";
	;;
	"BOTH")
		FASTQ_FILES="$SCRATCH_DIR/$SAMPLE/$SAMPLE.extendedFrags.fastq";
		FASTQ_FILES_SECOND_BATCH="$SCRATCH_DIR/$SAMPLE/$SAMPLE.notCombined_1.fastq $SCRATCH_DIR/$SAMPLE/$SAMPLE.notCombined_2.fastq";
	;;
esac

bwa mem -t $THREADS $GENOME_DIR $FASTQ_FILES > $SCRATCH_DIR/$SAMPLE/mapped.sam;
if [ "$FLASH" == "BOTH" ];
then
	bwa mem -t $THREADS $GENOME_DIR $FASTQ_FILES_SECOND_BATCH > $SCRATCH_DIR/$SAMPLE/mapped.notCombined.sam;
fi
rm *fastq;

samtools view -bS -@ $THREADS $SCRATCH_DIR/$SAMPLE/mapped.sam > $SCRATCH_DIR/$SAMPLE/mapped.bam;
rm -r $SCRATCH_DIR/$SAMPLE/mapped.sam;
samtools sort -@ $THREADS $SCRATCH_DIR/$SAMPLE/mapped.bam $SCRATCH_DIR/$SAMPLE/$SAMPLE.sorted;
#samtools index -b $SCRATCH_DIR/$SAMPLE/$SAMPLE.bam;
rm -r $SCRATCH_DIR/$SAMPLE/mapped.bam;

if [ "$FLASH" == "BOTH" ];
then
	#sort second sam file
	mv $SCRATCH_DIR/$SAMPLE/$SAMPLE.sorted.bam $SCRATCH_DIR/$SAMPLE/$SAMPLE.flash.bam;
	rm $SCRATCH_DIR/$SAMPLE/mapped.bam;
	samtools view -bS -@ $THREADS $SCRATCH_DIR/$SAMPLE/mapped.notCombined.sam > $SCRATCH_DIR/$SAMPLE/mapped.bam;
	rm $SCRATCH_DIR/$SAMPLE/mapped.notCombined.sam;
	samtools sort -@ $THREADS $SCRATCH_DIR/$SAMPLE/mapped.bam $SCRATCH_DIR/$SAMPLE/$SAMPLE.notflash;
	rm $SCRATCH_DIR/$SAMPLE/mapped.bam;
	#samtools index -b $SCRATCH_DIR/$SAMPLE/$SAMPLE.notflash.bam;
	#merge the 2 bam files
	samtools view -H $SCRATCH_DIR/$SAMPLE/$SAMPLE.notflash.bam | grep -v '^\@RG' > $SCRATCH_DIR/$SAMPLE/header.sam;
	samtools view -H $SCRATCH_DIR/$SAMPLE/$SAMPLE.notflash.bam | grep '^\@RG' | sort -u > $SCRATCH_DIR/$SAMPLE/rg.sam;
	samtools view -H $SCRATCH_DIR/$SAMPLE/$SAMPLE.flash.bam | grep '^\@RG' | sort -u >> $SCRATCH_DIR/$SAMPLE/rg.sam;
	cat $SCRATCH_DIR/$SAMPLE/rg.sam | sort -u >> $SCRATCH_DIR/$SAMPLE/header.sam;
	samtools merge -h $SCRATCH_DIR/$SAMPLE/header.sam $SCRATCH_DIR/$SAMPLE/merged.bam $SCRATCH_DIR/$SAMPLE/$SAMPLE.flash.bam $SCRATCH_DIR/$SAMPLE/$SAMPLE.notflash.bam;
	rm $SCRATCH_DIR/$SAMPLE/header.sam $SCRATCH_DIR/$SAMPLE/rg.sam;
	rm $SCRATCH_DIR/$SAMPLE/$SAMPLE.flash.bam;
	rm $SCRATCH_DIR/$SAMPLE/$SAMPLE.notflash.bam;
	samtools sort -@ $THREADS $SCRATCH_DIR/$SAMPLE/merged.bam $SCRATCH_DIR/$SAMPLE/$SAMPLE.sorted;
	rm $SCRATCH_DIR/$SAMPLE/merged.bam;
fi

#run picard tools for correcting readgroups...
picard AddOrReplaceReadGroups I=$SCRATCH_DIR/$SAMPLE/$SAMPLE.sorted.bam O=$SCRATCH_DIR/$SAMPLE/$SAMPLE.bam RGID=$SAMPLE RGLB=$SAMPLE RGPL=illumina RGSM=$SAMPLE RGPU=$SAMPLE RGCN=GenomicsCoreLeuven CREATE_INDEX=TRUE;

rm $SCRATCH_DIR/$SAMPLE/$SAMPLE.sorted.bam;

mkdir -p $OUTPUT_DIR;
cd $OUTPUT_DIR;
rsync -ahr $SCRATCH_DIR/$SAMPLE/$SAMPLE* .;
rm -rf $SCRATCH_DIR/$SAMPLE;



